### DISCUSSION TITLE
Push-pull: What if trains could back up?

### DISCUSSION THREAD

**Discussion by 2TallTyler** [OpenTTD/OpenTTD#12886] [Features]
# Done!

Implemented in https://github.com/OpenTTD/OpenTTD/pull/15379 and https://github.com/OpenTTD/OpenTTD/pull/15391 (among others).

---

## Motivation

Players want trains to not magically flip in stations, if there is a driving cab on the other end.

## Proposed implementation

1. Add a NewGRF vehicle flag to mark an unpowered wagon as having a cab, and able to lead a train. This allows it to take the first position in a depot, like an engine. Many trainsets do this with driving trailers, snowplows, etc., by coding it as a powered unit with a small amount of power.
2. Change behavior of train flipping. If the last wagon is a powered loco, is a dual-headed engine, or has the flag from step 1, the order of the wagons is reversed at the moment the train flips direction, so that it appears to not flip. 
   - We need to determine how to handle NewGRF trainsets using the position in the consist variables. If we don't change it, automatic dining cars would shift position. But if we change it to count backwards when the train is flipped, trains wouldn't know when to draw headlights.

## What about freight trains?

Players want freight trains to back up/propel/shove at reduced speed. I'm not sure how to handle the vehicle chain, but as far as the user interface I suggest an order flag to be set manually by the player, instead of doing it automatically or using an invisible trailing wagon with the cab flag. One of my greatest frustrations with the shunting patch was when a train took a path I wasn't expecting and got itself turned around: it would operate in reverse, clogging up my network, until I could figure out how to turn it back around. Using an order flag would require the player to tell the train to drive to its next order backwards, avoiding this confusion.

**Comment by PeterN** [comment]
This proposal seems to require a consist with an engine at each end which seem a restrictive and not really solving much.

Something like...

* Separate consist information (stored in the lead vehicle) from vehicles. (Someone was working on a patch for that.)
* Allow a consist to exist that doesn't start with a vehicle type that is flagged as an engine.
* Either: allow vehicles to move backwards, or allow articulated parts to be handled in reverse.

**Reply by reldred** [reply]
+1, I just want trains not to flip. I'll handle turning them around again elsewhere if needed.

**Comment by LC-Zorg** [comment]
> Players want trains to not magically flip in stations, if there is a driving cab on the other end.

I understand such players and honestly I like to see such realistic lines. That's why I voted for your Title Screen for no. 13. And that's why I consider your save [100% connections on FIRS 4 Steeltown](https://www.tt-forums.net/viewtopic.php?t=88716) to be one of the most enjoyable to watch of all I've seen. :) But the claim that players (i.e. all or the vast majority) don't want trains to magically rotate is not true. This isn't a drawback for everyone. Not everyone likes micromanagement and time-consuming creation of strange structures everywhere that would allow the train to rotate. Besides, in my opinion, trains traveling the entire long route backwards will look just as or maybe even more frivolous than magical rotation in place.

I probably won't be a user of this feature, but looking at it a bit from the outside, I don't think that specifying every time how the train should behave (sometimes realistically and other times not) is a good solution. The disadvantage of this solution would also be adding new interface elements, which for many, and I think for most players would be useless and would only confuse the interface. You could probably make these elements visible only when the function was enabled. But is this manual definition really that necessary?

I won't go into technical details and what is possible and what is not, but I think adding a simple option like "don't allow trains to rotate in place" would probably be a good solution. Trains would always behave the same way.

By the way, how would trains behave after entering the depot? Would they magically rotate or not? If not, what would be consistent, I think it would be worth adding depots earlier with the possibility of passing through them.

**Reply by Pixel-Tony** [reply]
It appears to me that although your points stand mostly valid, you've missed one of the main specifications and also the meaning of this whole discussion: we don't want to disable "magical" 180 degree turns, we want to enable trains to keep their initial formation *in case there's a driving trailer at the consists end*.

**Reply by Pixel-Tony** [reply]
And, imho, in case an approach to this problem is developed, we *might* implement a 180-turn restriction option (mainly because there will be, at that point, an existing approach to playing with it enabled).

**Comment by reldred** [comment]
"Players want trains to not magically flip in stations, if there is a driving cab on the other end."

Personally, I don't want trains to magically flip at all, I'll handle turning trains back around again somewhere else in the station complex/nearby area if they need to (things that aren't MU's or don't have driving cabs).

**Comment by Ragin1924** [comment]
Obviously, you know my take, I'd love for this to be a feature. I could imagine code wise it might have some difficulties, but the ability to have a flag that says 'this is a cab car, let it lead a train when reversed' would be insanely useful. Ideally I'd like to have the flag work across GRFs, so you could theoretically use cab cars from my set with locos from iron horse and vice versa.

**Reply by audigex** [reply]
I definitely prefer the idea of a `driving_car` flag

It would also be nice to allow us to set that same flag to 0 in order to allow things like slugs or B-units (which have power, but cannot be used to drive the train). Currently if it has power you have to do some `position_in_consist` type jankiness to 

No flag would presumably continue current behaviour (power = can drive if at the front, no power = can't drive) but the present of either a 1 (can drive from front or rear regardless of power, as long as the train has power somewhere) or 0 (adds power but can't drive) flag would modify that current "default" behaviour

**Comment by andythenorth** [comment]
"Consists" 😛

**Comment by audigex** [comment]
I don't think we should require an engine at either end - rather more useful would be a "driving car" flag that can be set. If there's a driving car on the back with that flag, draw the train as though it was heading in the other direction. The flag also allows us to not mess up sets that already do push-pull in their own way

Ideally I'd add (terminology from NML for simplicity), and let's assume this is all layman terms rather than code-specific to the game

- A "Driving Car" flag that can be applied to non-engine vehicles to enable this functionality
- A shunt speed" variable/callback that can be set to reduce speed when travelling backwards (if not set, the train's max speed is used)
- `position_in_consist_as_built` (same functionality as `position_in_consist` now) and `position_in_consist_as_travelling` (as you'd expect, the position from the "logical" front of the train) and equivalents for `_in_articulated_vehicle` and  `_from_reverse` etc etc. Leave the current ones as-is. I guess the _as_built is optional
- Some method for supplying a "reversed" sprite. Essentially a `default_reversed` callback, for assymetric vehicles or headlights/tail lights etc

Fundamentally I think neither the newGRF developer nor the player care what's happening behind the scenes... they just want the train to be drawn as if it was travelling in the other direction. It doesn't need to actually be flipped, just drawn as though it was

**Reply by Brickblock1** [reply]
Would shunt speed be used when there isn't a cab or when there is? 

I also believe that `default_reverse` would be redundant

**Reply by audigex** [reply]
Shunt speed would be used when the rear vehicle has the `driving_car` flag, and the `shunt_speed` var/callback is set (I feel like a callback would be nicer, so we can do things like vary the speed based on the length of the consist, but I'd take argument on it)

If there's no driving car flag on the rear car the unit can't reverse, presumably, so shunt speed is irrelevant. If the shunting speed is not set, the max speed remains the same as it was before reversing (max speed of the slowest wagon in the consist, I think?)

`default_reverse` I've not fully thought through here, I guess it might be possible to do the same thing with `default` and `vehicle_is_reversed` like we already do now with hacky push-pull implementations?

**Reply by felixprigge** [reply]
Generally speaking, I think an option where trains stop flipping magically and actually reverse would be a huge benefit to the game.
To your point that a train cant reverse without a driving cab at the end: 
I think it would probably break the game. When this train gets stuck somewhere, there is maybe no chance to get it back on track. Maybe it is even not possible to extend the rails in front, because of industry or a town. That is also not really realistic.
I would love to see instead that trains without a driving cab at the end can back up, but painfully slow. Because they need do be guided from an external person. Maybe at around 15 kph or so, whatever is enjoyable to play. This speed should be completely independent of the train speed and the power unit and can be a game constant. This way, no additional variable is required for the engines and it shouldn't mess wit other train sets.

**Comment by felixprigge** [comment]
Out of curiosity and pure excitement I started implementing a reversing train feature myself. Just to see how far I would get.
My first try was to change the order of the consist at revesering like 2TallTyler proposed. But this would also mean that the order in the train gui (which shows the loading of the train) is changed, also the order of the consist inside the depot is depending if the train drives in normaly or backwards. For a fright train you would have a free waggon at the first place, which needs some additional code correction to be allowed. I am sure you can overcome this problem but I think it could be implemented easier.

My implementation is still heavily in progress but this is what I came up with so far:
- Add a newgame difficulty setting: enable train reverse
- Add this setting in the vehicle pathfinding menu.
- Rename everything that means "front" as in FrontEngine to "first" to avoid later confusion
- Driving direction is toggled by already existing TOGGLE_REVERSE flag 
- Add functions to get the pointer to "front", "ahead of" and "behind" unit of a consist depending on driving direction.
- Don't swap units at reversing if the enable train reverse setting is true, but revert direction and GOING_UP and GIONG_DOWN bits.
- The first engine (displayed at first position in Depot and GUI) still holds all relevant information of the consist and is still the unit which is pointed to when referring to a train.
- Now in train_cmd.cpp add in all functions, which are position dependend, a pointer to the front unit (called f instead of v). Use this pointer if we evaluate TRACK, TILE or DIRECTION. Otherwise use first engine v for everything else, including DEST_TILE which is still stored in the first engine.

For building trains inside a depot nothing changes with this implementation, which is nice. You still need an engine at the first position to have a valid consist. 
When merging two trains inside a depot, one that drove in normally and one that drove in backwards, we would need to overwrite the TOGGLE_REVERSE bit of the new consist with the value of the new first engine. This maybe give a bit unexpected results but is a rare case, I guess.

This is so far the mechanic to enable the reverse. Of couse it also needs to be defined what happens with the speed when the train reverses. As others mentioned, I would go with a penalty speed if the front is a waggon for shunting and regular speed if the front is either an engine or a rear part of multihead. 

https://github.com/felixprigge/OpenTTD_reversing_trains

**Reply by nielsmh** [reply]
I was about to suggest that the "Enable train reverse" setting should be part of the old "Disallow train reversing in stations" (any station/end of line only) setting, but then I realize that they are not direct extensions of each other. The latter, old setting, that traces back to the difficulty settings of original Transport Tycoon, determines whether a train may _change driving direction_ at any station, and there might be situations where you want to allow that while still requiring that they don't magically flip around.

You should probably also change the text of the old setting, to clarify how the two are different.

**Reply by felixprigge** [reply]
This is a valid point to avoid confusion, in the settings as well as in the code. Maybe a native speaker and train enthusiast can give some hints on really clear wording for this reversing, backing up, shunting situation.

**Comment by bjornwarmedal** [comment]
For me personally there is a limit to the realism I want in the game, and this is past that. Terminus stations wouldn't work at all unless you had engines in each end. Some articulated trains already sort of solve this: if the back end of the train is a mirror of the front end the player won't see a difference when the train flips.

For me trains flipping in place is up there with infinitely many infinitely long trains fitting in a one-tile depot. It's unrealistic, but not allowing it would make the game a lot more complicated for no real benefit.

**Reply by bjornwarmedal** [reply]
Oh, yeah: let's say you could place a steam engine in each end, with one of them turned the other way, to not have to turn around in some manner. That would almost double the cost of a train in many cases.

**Reply by reldred** [reply]
then don't turn the feature on? I don't think anyone's suggesting this should be forced.

**Reply by bjornwarmedal** [reply]
That's always an option, yeah. The question is how many players would use it, in relation to how much work it would be to implement. And the game has a tonne of settings to navigate already; would this just disappear in the ocean of other settings and make it even less worth to implement? :shrug:

Of course, if someone feels strongly enough for it to implement it then by all means go ahead :) I don't mean to kill anyone's buzz

**Reply by 2TallTyler** [reply]
I develop OpenTTD for myself first, and I want this feature. If nobody uses it besides me, it'll still be worth my time investment. 😉

**Reply by felixprigge** [reply]
@2TallTyler That is also my thinking! Are you actively working on this feature?

**Reply by 2TallTyler** [reply]
I am not actively working on it, no. If someone tackled the tech debt of converting trains from linked lists to consist objects (head engines have a lot of weirdness to deal with), I would have an easier time getting started. I'm too busy with work at the moment to tackle both tech debt and this feature.

**Reply by felixprigge** [reply]
@2TallTyler so you mean the whole train shall be a single object with arrays of information about the vehicles of the consist instead of the linked list of objects? Sounds like a nice thing, but seems to me like a huge change of code. 
What major benefits would this have? 
(Apart from making it easier to change the code for people like me, who are more familiar with python than with c++ and pointers :D )

**Comment by zephyris** [comment]
There's a lot of nice ideas here, but a bit clouded by a NewGRF-first discussion. In my opinion, there are one or two game settings, then a few NewGRF ideas which build on it.

Settings:
1. Allow train reversing / Disallow train reversing.
Whether or not trains should 'magically flip' in stations.
2. Trains without a powered vehicle at the front: Travel at full speed / Travel at some fraction of their full speed / Can't move.
Could be generalised to a single percentage, how fast a train without a powered vehicle at the front can travel from 100 to 0%.

@2TallTyler mentioned an order flag - presumably allow/disallow flipping at this station(?). IMO this is too granular - the player probably either wants the behaviour globally, or doesn't want it.

NewGRF:
1. A flag meaning 'this counts as a powered vehicle' would allow driving cab behaviours. 
2. A callback to set speed (as a fraction of maximum) if this vehicle is leading would allow nice extra behaviours.
4. A variable for driving direction. This would allow position in consist interpretation to be flipped, to make the train appearance unchanged, but expose a new variable for NewGRF authors to eg. choose which end has headlights.

*edit*
My suggestion doesn't cover something requested by GRF authors:
Allow train reversing but without flipping the order of the vehicles, with the exception of articulated vehicles (and longer consists?):
Current behaviour: <-Engine- <-Tender- Wagon Wagon Wagon [flips to] Wagon Wagon Wagon -Tender-> -Engine->
Altered behaviour: ... [flips to] -Tender-> -Engine-> Wagon Wagon Wagon
This gives a 'runaround' behaviour.
IMO this is a subset of the 'disallow train reversing' case for setting 1, which goes on to behave nicely with the suggested extra NewGRF features.

**Reply by Brickblock1** [reply]
I think that trains with the ability to not flip should never flip wether or not trains can reverse without a cab.

**Reply by Brickblock1** [reply]
I don't think using fraction of max speed really makes sense in this. Other than that the proposed newgrf stuff should be all that is necessary

**Reply by felixprigge** [reply]
just my 2 cents to your second setting: 
I think a train that can't move at all would lead to serious frustrations. In case it gets lost, drives in a terminus station and gets stuck, there will be no chance to get it back on track. I don't think, that people would use this option. 

If there should be an option to set a speed of going reverse without a driving cab at the front, I suggest to set an absolute value rather than a fraction of engine speed. Because this speed is not technically limited by physics, but by how good the traffic controller can guide the otherwise blind train. Obviously, this speed should not exceed engine speed ;)

**Reply by zephyris** [reply]
The thing to watch out for is game behaviour not matching with a graphical override from a NewGRF. If players want flipped trains to have a different behaviour - ie. move slower or not able to move - then NewGRFs which override that graphical feedback would be confusing.

**Reply by zephyris** [reply]
@felixprigge Slow internet, didn't see your post! It would be annoying, but as long as the visual feedback is clear (ie. its obvious which way it is travelling) I think it's a  good gameplay challenge. Could always plonk down a depot to drive it into...

**Reply by felixprigge** [reply]
Haha, not if the town does not give you the permit to tear down a building to build one. But yes, this is details. And if the reversing speed  can be set, it may also can be reduced down to zero. At the users own risk :)

@Brickblock1 You mean trains with driving cabs or engines at both ends should never flip and and freight trains without one at the end should flip depending on the setting? This seems inconsequential and not intuitive to me. I would prefer original behavior for all or not flipping for all.

**Reply by Brickblock1** [reply]
That is what I mean it is the current behavior after all

**Reply by felixprigge** [reply]
One other thing that bothers me is the head light and tail light thing with NewGRF. This refers to your point 3 to your NewGRF options.
In case the first and last unit of a consist is actually mirrored except that the first one heas head lights and the last one has tail lights we could swap the sprites, yes. But only in this case. In all other cases with lights indicating driving direction, we would need additional sprites.
For example, if we have a non symmetric diesel fright train engine, which can go both ways, it would need 16 sprites. 
My question is, if it would be generally more consequent and versatile for the future if the NewGRF had to provide additional sprites if it uses head lights and tail lights? My opinion is that another flag saying you can switch this engine with the mirrored rear part in case the train reverses and if there is no rear part than live with head lights shining back is a cheap work around, makes the code complex and does not really statisfies me. 
Yes, NewGRF developers have to add sprites if the train set should work perfectly with the reversing setting. But it is only changing the color of the lights...

**Reply by reldred** [reply]
I think there should definitely be a 'default' behavior that can be achieved requiring no input from NewGRF's - this will be useful for default trains, playing with legacy .grf's that won't see any updates, or .grf's from authors that flat out don't care. 

It should be reasonable in this case that steam engines (maybe check the VFX particles?) operate with a penalty, diesel and electric with less of a penalty, and finally DMU/EMU or consists with a trailing vehicle with zero penalty. The two settings approach Zeph has proposed would fit the bill I think.

It wouldn't need to be perfect, if someone wants perfect then a .grf utilizing the feature to its fullest can fill that gap.

**Reply by felixprigge** [reply]
Yes, if the grfs do not provide 16 sprites, than just use the 8 that it provides (or even just 4 in case of symmetric engines without lights). 
If the authors don't care, than the reversing train will have the wrong lights. So it will be. But as this behavior is already an additional option players can choose to use, they can decide to live with it.

**Reply by PeterN** [reply]
GRFs should test the driving direction variable and provide 4/8 sprites like normal, if they want to.

**Reply by Brickblock1** [reply]
Using the current variable would potentially break current grfs

**Reply by PeterN** [reply]
I didn't write "current variable".

**Comment by frosch123** [comment]
I missed this discussion. I just want to link my 12 year old prototype for this:
https://github.com/OpenTTD/OpenTTD/compare/master...frosch123:OpenTTD:feature_driving_backwards
It seems similar to @felixprigge 's branch.

There is also a discussion about it here: https://www.tt-forums.net/viewtopic.php?t=64384
Main reason I dropped the branch was, that it only support DMU/EMU style reversing, but no steam engine shunting. (see linked tt-forums topic)

**Reply by felixprigge** [reply]
Hi frosch123, thank you for sharing your old attempt and the discussion. I had a quick look at it. I think you almost had it, too bad that you didn't finish it. Although, the "allow flipping with penalty" approach wouldn't be my personal preferred approach.

**Reply by 2TallTyler** [reply]
If anyone wants to play with this, I previously rebased this branch to bring the code up to date (at least, now it's 1 year behind master, instead of 12): https://github.com/2TallTyler/OpenTTD/tree/backwards

**Reply by michicc** [reply]
On the topic on stale stuff: https://github.com/michicc/OpenTTD/tree/consist

Neither finished, nor doing any push-pull, but some ideas for a coding basis.

**Comment by felixprigge** [comment]
One more equestion we didn't discuss yet is AI behavior with backwards driving freight trains at reduced speed. 
Should we always do the magical flip for AIs?

**Comment by andythenorth** [comment]
1. as others have added above, this needs a Consist entity to act as a container.  Consist will have benefits for other sidequests.
2. behaviour of reversed trains, what is the proper domain?
  - global on a map?
  - per player company?
  - per vehicle group?
  - per vehicle?
  - per order?

**Reply by felixprigge** [reply]
because backing up trains have an economical drawback compared to flipping ones and increase difficulty, I suggest the domain to be all human player companies.

**Comment by felixprigge** [comment]
Hi all,
end of January I thought I would be close to a progress that could be ready for test playing. Pathfinding, signal and crash behavior seemed to work so far. Then my son was born and was afk as you may can imagine :smiley:. Last week I found some time to have another look, and figured out that the depot itself is another big challenge, as trains do not turn around in depots like they do at the end of a rail. Also the depot gui should show, in which direction the train will be driving when leaving. And what happens, when multiple trains are merged, some of them drove backwards into the depot and some not. 

There is still lots of work to do and I would like to continue on this but I am not sure if I will be able to finish it.

**Reply by 2TallTyler** [reply]
I suggest that either the depot show the vehicles in the train in the order it will leave the depot (so if you want it to back out of the depot, put the engine at the back), or always have trains leave a depot facing forwards. When I was playing with the old shunting patch, trains left the depot forwards, which was helpful for resetting things after making a mistake and getting a train facing the wrong way.

**Reply by Pixel-Tony** [reply]
I too think it is expected behavior for trains to always leave depots facing forwards. You decide what loco to put and what to use for the tail cab, but after you do, you *intend* for it to have a "right" direction as it is now, front to back as left to right. After all, we want it not to switch directions. You can just (as Tyler said) put the loco on the back if you want to use another consist order.

**Reply by felixprigge** [reply]
Yes, this sounds like a good solution to me as well. But I now come to the point, that I think I chose the wrong strategy of implementation by leaving the consist order (by means of pointers) like it is, with the engine as fist entry and invoke the path finding and signalling functions not from the first consist unit but from the front unit depending on driving direction. 

The other strategy to actually change the order of the whole consist (and disabling all the checks if the first unit is an engine) now seems a little bit easier to implement and would lead to your suggested behavior inside the depot. It would also benefit from the already mentioned consist container class, which could then be the link to the vehicle ID and store power and speed. 

But the solution it is not clear about the graphical direction of the single units. If I build a train with wagons first and engine last (because I want it to drive out of the depot backwards) the graphical direction of the loco has to be be changed as well. But you cant just check if the loco is first or last unit and set the graphical direction accordingly, because maybe there are multiple locos in the consist. And they should point to the same direction. Not even thinking of dualhead units. This also is not really straight forward. :sweat_smile: 
Loving this discussion!

**Reply by zephyris** [reply]
IMO reversing train direction should be be an active decision - eg. a reverse train button, while in the depot, or a go to depot and reverse order.

**Reply by Rito13** [reply]
Depots don't seem to be big enough for trains to turn around. I would expect train to leave the depot in the opposite direction, as like it would leave the station. Especially when something like #8480 would have been merged.

**Reply by felixprigge** [reply]
Just want to give an update here. I took some time for continuing on this project the last days and am confident that I will be able to present a solution for test playing within the next two weeks.
Maybe it is best to include a setting like "Trains will leave depots always forward".  Then we can try it out and see, which one makes more fun to play.

**Comment by 2TallTyler** [comment]
Implemented in #15379 and #15391 (among others).
